modules/reindex/src/main/java/org/elasticsearch/index/reindex/StartReindexJobAction.java

```diff
@@ -0,0 +1,184 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex;
+
+import org.elasticsearch.action.Action;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.CompositeIndicesRequest;
+import org.elasticsearch.action.support.master.AcknowledgedRequest;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+
+public class StartReindexJobAction extends Action<StartReindexJobAction.Response> {
+
+    public static final StartReindexJobAction INSTANCE = new StartReindexJobAction();
+    // TODO: Name
+    public static final String NAME = "indices:admin/reindex/start_reindex";
+//    public static final String NAME = "cluster:admin/reindex/start_reindex";
+//    public static final String NAME = "indices:data/reindex/start_reindex";
+
+    private StartReindexJobAction() {
+        super(NAME);
+    }
+
+    @Override
+    public StartReindexJobAction.Response newResponse() {
+        return new StartReindexJobAction.Response();
+    }
+
+    public static class Request extends AcknowledgedRequest<Request> implements ToXContentObject, CompositeIndicesRequest {
+
+        private ReindexRequest reindexRequest;
+        private boolean waitForCompletion = false;
+
+        public Request() {
+
+        }
+
+        public Request(ReindexRequest reindexRequest) {
+            this.reindexRequest = reindexRequest;
+        }
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            reindexRequest = new ReindexRequest(in);
+            waitForCompletion = in.readBoolean();
+        }
+
+//        public static Request fromXContent(final XContentParser parser, final String id) throws IOException {
+//            Request request = new Request();
+//            request.setReindexJob(ReindexJob.fromXContent(parser));
+//            return request;
+//        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
```

Comment:
No `readFrom` please, use constructor with StreamInput only and make fields of class final